docs/sbft-formal-model/messages.dfy

```diff
@@ -73,13 +87,21 @@ module Messages {
                      | Prepare(view:ViewNum, seqID:SequenceID, operationWrapper:OperationWrapper)
                      | Commit(view:ViewNum, seqID:SequenceID, operationWrapper:OperationWrapper)
                      | ClientRequest(clientOp:ClientOperation)
-                     | ViewChangeMsg(newView:ViewNum, certificates:imap<SequenceID, PreparedCertificate>) // omitting last stable because we don't have checkpointing yet.
+                     | ViewChangeMsg(newView:ViewNum,
+                                     lastStableCheckpoint:SequenceID,
+                                     proofForLastStable:CheckpointsQuorum,
+                                     certificates:map<SequenceID, PreparedCertificate>)
                      | NewViewMsg(newView:ViewNum, vcMsgs:ViewChangeMsgsSelectedByPrimary) 
                      | CheckpointMsg(seqIDReached:SequenceID)
                      {
-                       predicate WF() {
-                         && (ViewChangeMsg? ==> Library.FullImap(certificates)) //TODO: rename TotalImap
+                       predicate valid(quorumSize:nat) {
+                         && (ViewChangeMsg? ==> proofForLastStable.valid(lastStableCheckpoint, quorumSize))
+                         && (NewViewMsg? ==> vcMsgs.valid(newView, quorumSize))
                        }
                      }
+  predicate CheckMessageValidity(msg:Message, quorumSize:nat) {
```

Comment:
Maybe `IsMessageValid` is a better name for a predicate?

Reply:
This is a temporary fix resolving a problem with Dafny's recursion expansion heuristics. Yesterday we had discussion with Oded Padon and he suggested a much better solution which I have added in the latest commit.

Reply:
Thanks for the clarification!